presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java

```diff
@@ -136,13 +148,42 @@ public VariableStatsEstimate visitCall(CallExpression call, Void context)
                 return value.accept(this, context);
             }
 
-            // value is not a constant but we can still propagate estimation through cast
+            // value is not a constant, but we can still propagate estimation through cast
             if (resolution.isCastFunction(call.getFunctionHandle())) {
                 return computeCastStatistics(call, context);
             }
+
+            return computeStatsViaAnnotations(call, context, functionMetadata);
+        }
+
+        private VariableStatsEstimate computeStatsViaAnnotations(CallExpression call, Void context, FunctionMetadata functionMetadata)
+        {
+            // casting session to FullConnectorSession is not ideal.
+            boolean isStatsPropagationEnabled =
+                    SystemSessionProperties.shouldEnableScalarFunctionStatsPropagation(((FullConnectorSession) session).getSession());
+
+            if (isStatsPropagationEnabled) {
+                if (functionMetadata.hasStatsHeader() && call.getFunctionHandle() instanceof BuiltInFunctionHandle) {
+                    Signature signature = ((BuiltInFunctionHandle) call.getFunctionHandle()).getSignature().canonicalization();
+                    Optional<ScalarStatsHeader> statsHeader = functionMetadata.getStatsHeader(signature);
+                    if (statsHeader.isPresent()) {
+                        return computeCallStatistics(call, context, statsHeader.get());
+                    }
+                }
+                else {
+                    log.debug("Stats not found for func: " + functionMetadata.getName() + " " + call);
```

Comment:
Use printf style log statements - `log.debug("Scalar stats annotation missing for function : %s %s", functionMetadata.getName(), call)`
Ideally, we avoid debug logs in Presto